rfc/rfc-0020-resources-update-after-success.md

```diff
@@ -0,0 +1,91 @@
+# Draft RFC 20 Update Resources Only After Success/Failure
+
+## Summary
+
+Cartographer is currently unable to determine when an input to a resource has reached success, failed, or is still
+processing. Making Carto aware of status indications is not sufficient, as frequent updates to a resource could keep it
+in a constant state of 'processing', even as it succeeds and fails on multiple successive inputs. Carto should wait
+until an input has resulted in success or failure before updating the resource with new input. 
+
+## Motivation
+
+Connecting an output of a resource to an input is necessary for establishing provenance. That is to state, "The app 
+currently running on the cluster is a result of commit X," it is necessary to tie a resource output to the input that
+produced it. Waiting for success/failure before update achieves this.
+
+Tieing inputs to outputs also allows Carto to harden supply chains to tampering. Currently, an enemy could update a
+resource. Carto will not read said resource without first updating it with the proper spec. But the resource may still
+produce the enemy output (followed by the correct output). Carto must be able to associate inputs to outputs if it is
+not to propagate the enemy output.
```

Comment:
If a malicious actor has the ability to update the resource, it can also write to its status with malicious values, you're sunk. This proposal can't change that.

Reply:
Yes, if the enemy is a controller that can overwrite the status of an object, we cannot observe or defend against that. But if the enemy is one with only the ability to change an object's spec, e.g. a user with `kubectl` access, we can defend against them and there is value in doing so.

Reply:
> But if the enemy is one with only the ability to change an object's spec

I don't understand how it would be possible to only have write access to a resource's spec and not also have access to its status using bog standard RBAC. Can you explain the API access patterns you're thinking of?

Reply:
You can't use kubectl to write to an object's status. (With the design docs archived and scheduled for deletion, there doesn't seem to be a canonical place where this is stated so I'll link to this [issue](https://github.com/kubernetes/kubectl/issues/564))